src/tools/rbd/action/Trash.cc

```diff
@@ -423,53 +423,79 @@ int execute_purge (const po::variables_map &vm,
     }
 
     json_spirit::mArray arr = json.get_obj()["pools"].get_array();
-    
+
     double pool_percent_used = 0;
     uint64_t pool_total_bytes = 0;
-    for(uint8_t i = 0; i < arr.size(); ++i) {
-      if(arr[i].get_obj()["name"] == pool_name) {
-        json_spirit::mObject stats =  arr[i].get_obj()["stats"].get_obj();
-        pool_percent_used = stats["percent_used"].get_real();
-        if(pool_percent_used <= threshold) {
-          std::cout << "rbd: pool usage is lower than or equal to "
-                    << (threshold*100)
-                    << "%" << endl;
-          std::cout << "Nothing to do" << std::endl;
-          return 0;
-        }
-            
-        pool_total_bytes = stats["max_avail"].get_uint64() +
-                           stats["bytes_used"].get_uint64();
-        break;
-      }
-    }
-    
+
+    std::map<std::string, std::vector<const char *>> datapools;
+
     std::sort(trash_entries.begin(), trash_entries.end(),
-      [](librbd::trash_image_info_t a, librbd::trash_image_info_t b) { 
+      [](librbd::trash_image_info_t a, librbd::trash_image_info_t b) {
         return a.deferment_end_time < b.deferment_end_time;
       }
     );
 
-    uint64_t bytes_to_free = 0;
-    auto bytes_threshold = (uint64_t)(pool_total_bytes * 
-                           (pool_percent_used - threshold));
-    
-    librbd::Image curr_img;
     for (const auto& entry : trash_entries) {
-      r = utils::open_image_by_id(io_ctx, entry.id, true, &curr_img);
+      librbd::Image image;
+      std::string data_pool;
+      r = utils::open_image_by_id(io_ctx, entry.id, true, &image);
       if(r < 0) continue;
+
+      int64_t data_pool_id = image.get_data_pool_id();
+      if (data_pool_id != io_ctx.get_id()) {
+        librados::Rados rados(io_ctx);
+        librados::IoCtx data_io_ctx;
+        r = rados.ioctx_create2(data_pool_id, data_io_ctx);
+        data_pool = data_io_ctx.get_pool_name();
+        datapools[data_pool].push_back(entry.id.c_str());
+      }
+      datapools[pool_name].push_back(entry.id.c_str());
```

Comment:
On second thought -- I guess my suggestion would double-count an image's space in both its data pool and base pool. Therefore, the disk usage check for determining which images to delete should only be made in the data pool. Therefore, perhaps this should just be the ```else``` branch.

Reply:
Okay. With your prev suggestion I reasoned that technically all images belong to the base pool anyway and should be fine if we delete (datapool) images from basepool (where datapool's threshold is what is required and basepool is higher). But not double-counting makes it easier to differentiate and the (trash purge) behavior more predictable. Will modify.

Reply:
Done.